lib/unifex/code_generator/base_type.ex

```diff
@@ -99,69 +101,88 @@ defmodule Unifex.CodeGenerator.BaseType do
   """
-  @spec generate_initialization(t, name :: atom, module) :: CodeGenerator.code_t()
-  def generate_initialization(type, name, code_generator) do
-    call(type, :generate_initialization, [name], code_generator)
+  @spec generate_initialization(t, name :: atom, module, ctx :: map) :: CodeGenerator.code_t()
+  def generate_initialization(type, name, code_generator, ctx) do
+    call(type, :generate_initialization, [name], code_generator, ctx)
   end
 
   @doc """
   Generates an destrucition of variable content. Should be paired with `generate_initialization/1`
 
   Returns an empty string if the type does not provide destructor
   """
-  @spec generate_destruction(t, name :: atom, module) :: CodeGenerator.code_t()
-  def generate_destruction(type, name, code_generator) do
-    call(type, :generate_destruction, [name], code_generator)
+  @spec generate_destruction(t, name :: atom, module, ctx :: map) :: CodeGenerator.code_t()
+  def generate_destruction(type, name, code_generator, ctx) do
+    call(type, :generate_destruction, [name], code_generator, ctx)
   end
 
   @doc """
   Generates parsing of UNIFEX_TERM `argument` into the native variable
   """
-  def generate_arg_parse(type, name, argument, postproc_fun \\ & &1, code_generator) do
+  def generate_arg_parse(type, name, argument, postproc_fun \\ & &1, code_generator, ctx) do
+    ctx =
+      ctx
+      |> Map.put(:postproc_fun, postproc_fun)
+
     call(
       type,
       :generate_arg_parse,
       [argument, name],
       code_generator,
-      %{postproc_fun: postproc_fun}
+      ctx
     )
     |> postproc_fun.()
   end
 
-  def generate_arg_name(type, name, code_generator) do
-    generate_native_type(type, code_generator)
+  def generate_arg_name(type, name, code_generator, ctx) do
+    generate_native_type(type, code_generator, ctx)
     |> Bunch.listify()
     |> Enum.map(fn
       {_type, suffix} -> ~g<#{name}#{suffix}>
       _type -> ~g<#{name}>
     end)
   end
 
-  def generate_native_type(type, mode \\ :default, code_generator) do
-    call(type, :generate_native_type, [], code_generator, %{mode: mode})
+  def generate_native_type(type, mode \\ :default, code_generator, ctx) do
+    ctx =
+      ctx
+      |> Map.put(:mode, mode)
+
+    call(type, :generate_native_type, [], code_generator, %{ctx | mode: mode})
   end
 
-  defp call(full_type, callback, args, code_generator, ctx \\ %{}) do
+  defp call(full_type, callback, args, code_generator, ctx) do
     {type, subtype} =
       case full_type do
         {type, subtype} -> {type, subtype}
         type -> {type, nil}
       end
 
+    ctx =
+      with %{customized_types: %{^type => type_spec}} <- ctx do
+        ctx
+        |> Map.put(:type_spec, type_spec)
+      else
+        _ -> ctx
+      end
+
     module =
-      Module.concat(Unifex.CodeGenerator.BaseTypes, type |> to_string() |> String.capitalize())
+      with %{customized_types: %{^type => _type_spec}} <- ctx do
```

Comment:
If we rename `customized_types` in ctx to e.g `struct_types` then we can do 
```elixir
if Map.has_key(ctx.struct_types, type), do: .., else: ..
```
which could be more intuitive.

Reply:
After adding to the context informations about new customized types, other than structures, this place in code would require change again

Reply:
Yeah I know but at this moment I can see only structs and enums. Don't think we would have more. If we don't want to change this I would rename `customized_types` to `user_types`?